{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 1: Controls\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Connection info needed in SQLAlchemy format, example:\n",
      "               postgresql://username:password@hostname/dbname\n",
      "               or an existing connection: dict_keys([])\n",
      "No module named 'pymysql'\n",
      "Connection info needed in SQLAlchemy format, example:\n",
      "               postgresql://username:password@hostname/dbname\n",
      "               or an existing connection: dict_keys([])\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "germplasm = open ('Germplasm.tsv', 'r')\n",
    "genelocus = open ('LocusGene.tsv', 'r')\n",
    "\n",
    "germplasm.seek(0)\n",
    "\n",
    " \n",
    "for line in germplasm.readlines():\n",
    "\n",
    "    matchObj = re.search(r'AT\\dG\\d+', line)\n",
    "\n",
    "    if at:\n",
    "\n",
    "        name1 = print(matchObj(1))\n",
    "\n",
    "    else:\n",
    "\n",
    "        pass\n",
    "\n",
    "germplasm.close()\n",
    "\n",
    "for line in genelocus.readlines():\n",
    "\n",
    "    matchObj = re.search(r'AT\\dG\\d+', line)\n",
    "\n",
    "    if at:\n",
    "\n",
    "        name2 = print(matchObj(1))\n",
    "\n",
    "    else:\n",
    "\n",
    "        pass\n",
    "\n",
    "genelocus.close()\n",
    "\n",
    "if name1==name2:\n",
    "\n",
    "            print (\"same\")\n",
    "\n",
    "         else:\n",
    "\n",
    "            print (“not found”)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 2: Design and create the database.¶\n",
    "It should have two tables - one for each of the two data files.\n",
    "The two tables should be linked in a 1:1 relationship\n",
    "you may use either sqlMagic or pymysql to build the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Line magic function `%sql` not found.\n"
     ]
    }
   ],
   "source": [
    "%sql create database germplasm;\n",
    "%sql use  germplasm;\n",
    "%sql show tables;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql CREATE TABLE germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(10) NOT NULL, Germplasm VARCHAR(10) NOT NULL, Phenotype VARCHAR(100) NOT NULL, Pubmed INTEGER NOT NULL);\n",
    "%sql DESCRIBE germplasm;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql CREATE TABLE locusgene (id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(10) NOT NULL, Gene VARCHAR(10) NOT NULL, ProteinLenght INTEGER NOT NULL);\n",
    "%sql DESCRIBE locusgene;\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql show tables;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database. There are a variety of strategies to accomplish this. I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasm',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True )\n",
    "germplasm = open (\"Germplasm.tsv\", \"r\")\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        germplasm.seek(0)\n",
    "        discard = germplasm.readline()\n",
    "        for line in germplasm.readlines():\n",
    "            line = line.strip() \n",
    "            if line:\n",
    "                Locus, Germplasm, Phenotype, Pubmed = line.split(\"\\t\")\n",
    "                sql = \"INSERT INTO germplasm (Locus, Germplasm, Phenotype, Pubmed) VALUES\n",
    "   ('%s','%s','%s','%s')\" % (Locus, Germplasm, Phenotype, Pubmed)\n",
    "                cursor.execute(sql)\n",
    "           else:\n",
    "             continue\n",
    "      \n",
    "        germplasm.close()\n",
    "\n",
    "finally:\n",
    "    connection.close()\n",
    "\n",
    " \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 4: Create reports, written to a file¶\n",
    "Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule!\n",
    "\n",
    "All reports should be written to the same file. You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM locusgene, germplasm WHERE \\\n",
    "     locusgene.id = germplasm.id;"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
